### DISCUSSION TITLE
Async actions batch (transactions)

### DISCUSSION THREAD

**Discussion by nbabanov** [mobxjs/mobx#3144] [Q&A] [answered]
Hello everyone! I need some help. I've read the https://mobx.js.org/actions.html#asynchronous-actions page, but it does not show a similar case like mine.

**What is happening?**
I have an action which triggers a couple of other actions, which are all async and modify the state at different times. This results in multiple observer notification instead of only one after all changes have been done.

**Why is this a problem?**
My UI is rerendered a lot of times due to the state being updated piece by piece. This creates visual glitches and makes the UX subpar.

**What is the ideal solution?**
I need to batch some async actions, but I don't see a way to do this. Look at the code below. I want to `update the MobX state` / `Notify the observers` only after `asyncAction` has resolved.

```js
const obj = {
  asyncAction = mobx.action(async () => {
    await functionThatChangesStateAfterResult();
    await anotherFunctionThatChangesStateAfterResult();
  });
}
```

After looking at the MobX internals I see that `executeAction` does not wait for the function to complete (if it's async) and synchronously calls `_endAction`. I want to have a transaction that wraps all async actions and then notifies the observers.

Thank you for your time!

**Comment by urugator** [comment, marked as answer]
There is nothing like async transaction. Create a local temporal object/variable that is modified by the individual async tasks and write the result to actual observables at the end of async flow.

**Reply by nbabanov** [reply]
Thank you, I haven't thought about this! I'll go this way.

**Comment by mweststrate** [comment]
Or capture the writes in a thunk to postpone them and run them at the end.

**Reply by nbabanov** [reply]
Thank you! In essence your answer is the same as @urugator . I'll see how I'll implement it, because it would be a change to a numerous places in one of our core APIs.
